<a href="https://colab.research.google.com/github/buttercutter/RNA-raiders/blob/main/RFGB.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [13]:
!pip install matplotlib scanpy scipy

Looking in indexes: https://pypi.org/simple, https://us-python.pkg.dev/colab-wheels/public/simple/


In [14]:
from sklearn.tree import DecisionTreeRegressor, DecisionTreeClassifier
from sklearn.ensemble import RandomForestRegressor
from sklearn.multioutput import MultiOutputRegressor
from sklearn.datasets import load_boston, load_breast_cancer
from sklearn.model_selection import train_test_split
from sklearn.metrics import accuracy_score, mean_absolute_error
from collections import Counter
import numpy as np
import pandas as pd
np.set_printoptions(precision=3)
import jax
import jax.numpy as jnp
from tqdm import tqdm 
import pickle as pkl
from typing import Tuple, Union
import scipy

In [15]:
import scanpy as sc

from warnings import filterwarnings
filterwarnings('ignore')

sc.settings.verbosity = 3
sc.logging.print_header()
sc.settings.set_figure_params(dpi=80, facecolor='white')

scanpy==1.9.1 anndata==0.8.0 umap==0.5.3 numpy==1.21.6 scipy==1.7.3 pandas==1.3.5 scikit-learn==1.0.2 statsmodels==0.12.2 pynndescent==0.5.8


In [16]:
import seaborn as sns
import matplotlib.pyplot as plt

# 1.1 Datasets

In [17]:
!wget https://saturn-public-data.s3.us-east-2.amazonaws.com/cancer-immunotherapy-challenge/data/sc_training.h5ad

--2023-01-14 13:12:02--  https://saturn-public-data.s3.us-east-2.amazonaws.com/cancer-immunotherapy-challenge/data/sc_training.h5ad
Resolving saturn-public-data.s3.us-east-2.amazonaws.com (saturn-public-data.s3.us-east-2.amazonaws.com)... 52.219.84.224
Connecting to saturn-public-data.s3.us-east-2.amazonaws.com (saturn-public-data.s3.us-east-2.amazonaws.com)|52.219.84.224|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: 1291266853 (1.2G) [binary/octet-stream]
Saving to: ‘sc_training.h5ad.3’

sc_training.h5ad.3  100%[===================>]   1.20G  30.0MB/s    in 42s     

2023-01-14 13:12:45 (29.0 MB/s) - ‘sc_training.h5ad.3’ saved [1291266853/1291266853]



In [18]:
!wget https://saturn-public-data.s3.us-east-2.amazonaws.com/cancer-immunotherapy-challenge/code/sc_training_visualization.ipynb

--2023-01-14 13:12:46--  https://saturn-public-data.s3.us-east-2.amazonaws.com/cancer-immunotherapy-challenge/code/sc_training_visualization.ipynb
Resolving saturn-public-data.s3.us-east-2.amazonaws.com (saturn-public-data.s3.us-east-2.amazonaws.com)... 52.219.108.194
Connecting to saturn-public-data.s3.us-east-2.amazonaws.com (saturn-public-data.s3.us-east-2.amazonaws.com)|52.219.108.194|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: 446576 (436K) [binary/octet-stream]
Saving to: ‘sc_training_visualization.ipynb.3’

sc_training_visuali 100%[===================>] 436.11K  1.10MB/s    in 0.4s    

2023-01-14 13:12:46 (1.10 MB/s) - ‘sc_training_visualization.ipynb.3’ saved [446576/446576]



In [19]:
!wget https://saturn-public-data.s3.us-east-2.amazonaws.com/cancer-immunotherapy-challenge/data/clone_information.csv

--2023-01-14 13:12:46--  https://saturn-public-data.s3.us-east-2.amazonaws.com/cancer-immunotherapy-challenge/data/clone_information.csv
Resolving saturn-public-data.s3.us-east-2.amazonaws.com (saturn-public-data.s3.us-east-2.amazonaws.com)... 52.219.108.194
Connecting to saturn-public-data.s3.us-east-2.amazonaws.com (saturn-public-data.s3.us-east-2.amazonaws.com)|52.219.108.194|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: 1542596 (1.5M) [text/csv]
Saving to: ‘clone_information.csv.3’

clone_information.c 100%[===================>]   1.47M  2.53MB/s    in 0.6s    

2023-01-14 13:12:48 (2.53 MB/s) - ‘clone_information.csv.3’ saved [1542596/1542596]



In [20]:
!wget https://saturn-public-data.s3.us-east-2.amazonaws.com/cancer-immunotherapy-challenge/data/guide_abundance.csv

--2023-01-14 13:12:48--  https://saturn-public-data.s3.us-east-2.amazonaws.com/cancer-immunotherapy-challenge/data/guide_abundance.csv
Resolving saturn-public-data.s3.us-east-2.amazonaws.com (saturn-public-data.s3.us-east-2.amazonaws.com)... 52.219.108.194
Connecting to saturn-public-data.s3.us-east-2.amazonaws.com (saturn-public-data.s3.us-east-2.amazonaws.com)|52.219.108.194|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: 4452 (4.3K) [text/csv]
Saving to: ‘guide_abundance.csv.3’

guide_abundance.csv 100%[===================>]   4.35K  --.-KB/s    in 0s      

2023-01-14 13:12:48 (252 MB/s) - ‘guide_abundance.csv.3’ saved [4452/4452]



In [21]:
!wget https://saturn-public-data.s3.us-east-2.amazonaws.com/cancer-immunotherapy-challenge/data/scRNA_ATAC.h5

--2023-01-14 13:12:48--  https://saturn-public-data.s3.us-east-2.amazonaws.com/cancer-immunotherapy-challenge/data/scRNA_ATAC.h5
Resolving saturn-public-data.s3.us-east-2.amazonaws.com (saturn-public-data.s3.us-east-2.amazonaws.com)... 52.219.108.194
Connecting to saturn-public-data.s3.us-east-2.amazonaws.com (saturn-public-data.s3.us-east-2.amazonaws.com)|52.219.108.194|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: 87606852 (84M) [application/x-hdf5]
Saving to: ‘scRNA_ATAC.h5.3’

scRNA_ATAC.h5.3     100%[===================>]  83.55M  24.1MB/s    in 3.5s    

2023-01-14 13:12:52 (24.1 MB/s) - ‘scRNA_ATAC.h5.3’ saved [87606852/87606852]



For *read_h5ad()*, see [scanpy documentation](https://scanpy.readthedocs.io/en/stable/generated/scanpy.read_h5ad.html)

In [22]:
adata = sc.read_h5ad('./sc_training.h5ad.2')
adata

AnnData object with n_obs × n_vars = 28697 × 15077
    obs: 'gRNA_maxID', 'state', 'condition', 'lane'
    layers: 'rawcounts'

For *adata* object type, see [AnnData documentation](https://anndata.readthedocs.io/en/stable/generated/anndata.AnnData.html#anndata.AnnData)

In [23]:
adata.obs

gRNA_maxID               state  \
053l1_AAACCTGAGATGTCGG-1  ONE-NON-GENE-SITE-7  terminal exhausted   
053l1_AAACCTGAGCAACGGT-1               Tox2-3            effector   
053l1_AAACCTGAGTACGACG-1               Tpt1-2            effector   
053l1_AAACCTGAGTCGTTTG-1               Tox2-3  terminal exhausted   
053l1_AAACCTGAGTGAAGAG-1               Tcf7-2            effector   
...                                       ...                 ...   
053l4_TTTGTCATCAGGTTCA-1               Tox2-3               other   
053l4_TTTGTCATCAGTGTTG-1               Dvl2-3             cycling   
053l4_TTTGTCATCCTCGCAT-1               Zeb2-2             cycling   
053l4_TTTGTCATCTTCAACT-1               Sox4-3             cycling   
053l4_TTTGTCATCTTTCCTC-1               Dvl2-3             cycling   

                            condition   lane  
053l1_AAACCTGAGATGTCGG-1  Unperturbed  lane1  
053l1_AAACCTGAGCAACGGT-1         Tox2  lane1  
053l1_AAACCTGAGTACGACG-1         Tpt1  lane1  
053l1_AAACCTGAGTCGTTTG-1         Tox2  lane1  
053l1_AAACCTGAGTGAAGAG-1         Tcf7  lane1  
...                               ...    ...  
053l4_TTTGTCATCAGGTTCA-1         Tox2  lane4  
053l4_TTTGTCATCAGTGTTG-1         Dvl2  lane4  
053l4_TTTGTCATCCTCGCAT-1         Zeb2  lane4  
053l4_TTTGTCATCTTCAACT-1         Sox4  lane4  
053l4_TTTGTCATCTTTCCTC-1         Dvl2  lane4  

[28697 rows x 4 columns]

# 1.1 Preprocess data

In [24]:
expression_data = adata.X
expression_data_sub = expression_data[:10, :10]
expression_data_sub.toarray()

array([[0.512, 0.   , 0.   , 0.   , 0.   , 0.   , 0.   , 0.   , 0.   ,
        0.   ],
       [0.484, 0.484, 0.809, 0.   , 0.   , 0.   , 0.   , 0.   , 0.484,
        0.   ],
       [0.   , 0.   , 0.   , 0.   , 0.   , 0.   , 0.694, 0.   , 0.   ,
        0.   ],
       [0.   , 0.   , 0.   , 0.   , 0.   , 0.   , 0.   , 0.   , 0.   ,
        0.   ],
       [0.   , 0.   , 0.   , 0.   , 0.   , 0.   , 0.   , 0.   , 0.   ,
        0.   ],
       [0.   , 1.089, 0.686, 0.686, 0.   , 0.   , 0.   , 0.   , 0.   ,
        0.   ],
       [0.   , 0.193, 0.493, 0.   , 0.   , 0.   , 0.193, 0.   , 0.724,
        0.   ],
       [0.   , 0.   , 0.   , 0.   , 0.   , 0.   , 0.   , 0.   , 0.   ,
        0.   ],
       [0.292, 0.292, 0.292, 0.292, 0.292, 0.   , 0.   , 0.   , 0.292,
        0.   ],
       [0.412, 0.704, 0.   , 0.   , 0.   , 0.   , 0.   , 0.   , 0.   ,
        0.   ]], dtype=float32)

Each row in this matrix represents a cell type (28,697 in total) and the column represents 15,077 genes. The elements indicate the expression levels. 

In [25]:
obs = adata.obs
genes = adata.var.index.tolist()
expression_data_csc = expression_data.tocsc()

test_genes = ['Ets1', 'Fosb', 'Mafk', 'Stat3']
validation_genes = ['Aqr', 'Bach2', 'Bhlhe40']

test_indices = [genes.index(x) for x in test_genes if x in genes]
validation_indices = [genes.index(x) for x in validation_genes if x in genes]

X_test = expression_data_csc[:, test_indices]
X_test = X_test.tocsr()
X_val = expression_data_csc[:, validation_indices]
X_val = X_val.tocsr()

all_indices = jnp.arange(expression_data_csc.shape[1])
rm_indices = test_indices + validation_indices
rm_indices.sort()
keep_indices = jnp.setdiff1d(jnp.array(all_indices), jnp.array(rm_indices))

expression_data_csc = expression_data_csc[:, keep_indices]
X_train = expression_data_csc.tocsr()

obs_cond_idx = obs.set_index('condition')

exclude_list = ['Unperturbed', 'Fzd1', 'P2rx7']
dropped_obs = obs_cond_idx.drop(index=exclude_list)

removed_indices = obs_cond_idx.index.isin(exclude_list)

X_train, X_test, X_val = X_train[~removed_indices], X_test[~removed_indices],\
                          X_val[~removed_indices]
X_train, X_test, X_val = X_train.toarray(), X_test.toarray(), X_val.toarray()

Now we have our train, test and validation samples. 



In [26]:
states = ['progenitor', 'effector', 'terminal exhausted', 'cycling', 'other']
ground_truth = {}

for gene in genes:
  state_frequency = {state: 0 for state in states}
  rows = obs[obs['condition'] == gene]

  if not rows.empty:  
    for index, row in rows.iterrows():
        state = row['state']
        if state not in states:
          state_frequency['other'] += 1
        else:
          state_frequency[state] += 1
    
    total_count = sum(state_frequency.values())
    for state, count in state_frequency.items():
        state_frequency[state] = count / total_count
    
    ground_truth[gene] = list(state_frequency.values())

Now we have a dictionary that stores the ground truth, i.e. the cell state frequencies, for each gene. Note that each list in this dictionary sums to 1, as required. We have 64 ground truth vectors because of the 66 knockout experiments that are used for training purposes, 2 did not pass quality control (see: [challenge 1 document](https://drive.google.com/file/d/1rR5oIhETmyVu6Uo5BsWjIokdBZctKzCE/view)). 

Next up, we'll have to generate our y_data and filter the unperturbed, and genes that didn't pass the quality control from our training and testing data.

In [45]:
y_data_dict = ground_truth.copy()
y_data = pd.DataFrame.from_dict(y_data_dict, orient='index', 
                                         columns=['progenitor', 'effector', 
                                         'terminal exhausted', 'cycling', 
                                         'other'])
y_data_expand = pd.DataFrame(columns=y_data.columns, index=dropped_obs.index)
idx = 0
for condition, row in dropped_obs.iterrows():
    gene = condition
    state_data = y_data.loc[gene]
    y_data_expand.iloc[idx] = state_data
    idx += 1

y_train = y_data_expand
y_train

progenitor  effector terminal exhausted   cycling     other
condition                                                            
Tox2        0.017309  0.057004           0.367874   0.53081  0.027002
Tpt1            0.44      0.16               0.12      0.28       0.0
Tox2        0.017309  0.057004           0.367874   0.53081  0.027002
Tcf7        0.104377  0.292929           0.313131  0.276094  0.013468
Tox2        0.017309  0.057004           0.367874   0.53081  0.027002
...              ...       ...                ...       ...       ...
Tox2        0.017309  0.057004           0.367874   0.53081  0.027002
Dvl2         0.02439  0.115509           0.304188  0.549471  0.006443
Zeb2        0.017483  0.115385           0.282051  0.439394  0.145688
Sox4        0.069697  0.112121           0.290909  0.512121  0.015152
Dvl2         0.02439  0.115509           0.304188  0.549471  0.006443

[23411 rows x 5 columns]

Note that we do not have y_test and y_val, as these will be used to evaluate the performance of our model in the competition. 

# 2. Random Forest

In [28]:
# class RandomForest:
    
#     def __init__(self, n_estimators:int=100, subsample:float=0.1, regression:bool=True, **kwargs)->None:
#         self.n_estimators = n_estimators
#         self.estimators = []
#         self.subsample = subsample
      
#         for _ in range(self.n_estimators):
#             if regression:
#                 self.estimators.append(DecisionTreeRegressor(**kwargs))
#             else:
#                 self.estimators.append(DecisionTreeClassifier(**kwargs))
            
               
#     def bootstrap_sample(self, X:np.array, y:np.array)->Tuple[np.array,...]:
#         n_samples = X.shape[0]
#         idxs = np.random.choice(n_samples, size=int(self.subsample*n_samples), replace=True)
#         return X[idxs], y[idxs]
    
#     def fit(self, X:np.array, y:np.array)->None:
#         for estimator in tqdm(self.estimators):
#             X_sample, y_sample = self.bootstrap_sample(X, y)
#             estimator.fit(X_sample, y_sample)

        
#     def predict(self, X:np.array)->np.array:
#         preds = np.array([estimator.predict(X) for estimator in self.estimators])
#         preds = np.swapaxes(preds, 0, 1)
#         #return np.array([self._most_common_pred(pred) for pred in preds])
#         return np.apply_along_axis(lambda x: np.bincount(x).argmax(), 1, preds)        
        
#     def _most_common_pred(self, y:np.array)->Union[float, int]:
#         #counts = np.bincount(y.ravel())
#         #return np.argmax(counts)

#         return np.mode(y, axis=None)[0][0]
#         #return scipy.stats.mode(y, axis=None)[0][0]
#         #return Counter(y.tolist()).most_common(1)[0][0]
#         #return Counter(y).most_common(1)[0][0]
#         #return scipy.stats.mode(y, axis=None)[0][0]

## 3.1. Random Forest Regression


In [41]:
rfr = RandomForestRegressor(n_estimators=100, random_state=42, warm_start=True, 
                            n_jobs=-1)

In [30]:
X = X_train
X.shape

(23411, 15070)

In [46]:
y_train = np.array(y_train)
# y_train = y_train.astype(np.int)
y = y_train

y

array([[0.017309023771059313, 0.05700438495268867, 0.36787445188091394,
        0.5308100623124856, 0.027002077082852526],
       [0.44, 0.16, 0.12, 0.28, 0.0],
       [0.017309023771059313, 0.05700438495268867, 0.36787445188091394,
        0.5308100623124856, 0.027002077082852526],
       ...,
       [0.017482517482517484, 0.11538461538461539, 0.28205128205128205,
        0.4393939393939394, 0.1456876456876457],
       [0.0696969696969697, 0.11212121212121212, 0.2909090909090909,
        0.5121212121212121, 0.015151515151515152],
       [0.024390243902439025, 0.1155085135757018, 0.3041877588587207,
        0.5494707777266452, 0.006442705936493327]], dtype=object)

In [47]:
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.4, 
                                                    random_state=42)

In [48]:
print('Number of estimators fitted:\n')
for i in tqdm(range(100)):
    # rfr.set_params(n_estimators=i+1)
    rfr.fit(X_train, y_train)

Number of estimators fitted:



100%|██████████| 100/100 [02:36<00:00,  1.57s/it]


In [49]:
y_pred = rfr.predict(X_train)
y_pred

array([[0.004, 0.015, 0.138, 0.171, 0.006],
       [0.011, 0.094, 0.092, 0.132, 0.005],
       [0.011, 0.039, 0.112, 0.169, 0.002],
       ...,
       [0.004, 0.015, 0.138, 0.171, 0.006],
       [0.1  , 0.03 , 0.082, 0.118, 0.003],
       [0.014, 0.103, 0.089, 0.122, 0.006]])

In [50]:
y_pred.shape

(14046, 5)

In [51]:
print(mean_absolute_error(y_test, rfr.predict(X_test)).round(3))

0.138


___

# 4. Gradient Boosting

In [ ]:
# Loss functions
def MSE(y_true:jnp.array, y_pred:jnp.array):
    return jnp.mean(jnp.sum(jnp.square(y_true-y_pred)))

def CrossEntropy(y_true:jnp.array, y_proba:jnp.array):
    y_proba = jnp.clip(y_proba, 1e-5, 1 - 1e-5)
    return jnp.sum(- y_true * jnp.log(y_proba) - (1 - y_true) * jnp.log(1 - y_proba))

In [ ]:
class GradientBoosting:

    def __init__(self, n_estimators:int=100, learning_rate:float=.1, regression:bool=True, **kwargs):
        self.n_estimators = n_estimators
        self.learning_rate = learning_rate
        self.regression = regression
        self.loss = MSE if self.regression else CrossEntropy

        self.estimators = []
        for _ in range(self.n_estimators):
                self.estimators.append(DecisionTreeRegressor(**kwargs))

    def fit(self, X:np.array, y:np.array):
        y_pred = np.full(np.shape(y), np.mean(y))
        for i, estimator in enumerate(self.estimators):
            gradient = jax.grad(self.loss, argnums=1)(y.astype(np.float32), y_pred.astype(np.float32))
            self.estimators[i].fit(X, gradient)
            update = self.estimators[i].predict(X)
            y_pred -= (self.learning_rate * update)

    def predict(self, X:np.array):
        y_pred = np.zeros(X.shape[0], dtype=np.float32)
        for estimator in self.estimators:
            y_pred -= (self.learning_rate * estimator.predict(X))
    
        if not self.regression:
            return np.where(1/(1 + np.exp(-y_pred))>.5, 1, 0)
        return y_pred

## 4.1 Gradient Boosting Regressor

In [ ]:
gbr = GradientBoosting(regression=True)

In [ ]:
gbr.fit(X_train_r, y_train_r)

NameError: ignored

In [ ]:
print(mean_absolute_error(gbr.predict(X_test_r), y_test_r).round(3))

## 4.2 Gradient Boosting Classifier

In [ ]:
gbc = GradientBoosting(regression=False)

In [ ]:
gbc.fit(X_train_c, y_train_c)

In [ ]:
print(accuracy_score(gbc.predict(X_test_c), y_test_c).round(3))